### ISSUE TITLE
Support webpack.config.js exporting functions and promises

### ISSUE THREAD

**Issue by third774** [cloudflare/wrangler#323]
According to the [webpack docs](https://webpack.js.org/configuration/configuration-types/) `webpack.config.js` should also be able to export a function that returns a config object, a promise that resolves to a config object, or a function that returns a promise of a config object.

I've found the function pattern to be useful for being able to easily compose various flags when doing local development with a `webpack.config.js` that looks like this:

```js
require("dotenv").config();

const webpack = require("webpack");
const BundleAnalyzerPlugin = require("webpack-bundle-analyzer")
  .BundleAnalyzerPlugin;

const DEFAULTS = {
  runtime: "webworker"
};

module.exports = (env = DEFAULTS, argv = { mode: "production" }) => {
  console.log("cliEnv", env);
  const { mode } = options;
  const mergedEnv = {
    ...DEFAULTS,
    ...env
  };
  return {
    mode,
    entry: "./src/index.ts",
    output: {
      path: __dirname,
      filename: "index.js"
    },
    module: {
      rules: [{ test: /\.(ts|js)$/, use: "babel-loader" }]
    },
    resolve: {
      extensions: [".ts", ".js"]
    },
    target: mergedEnv.runtime,
    externals: [],
    plugins: [
      mergedEnv.analyze && new BundleAnalyzerPlugin(),
      new webpack.EnvironmentPlugin({
        RUNTIME: mergedEnv.runtime,
        OMDB_API_KEY: process.env.OMDB_API_KEY,
        SPREADSHEET_ID: process.env.SPREADSHEET_ID
      })
    ].filter(Boolean)
  };
};
```

And a `package.json` that looks like this:

```json
{
  "name": "movie-night-data-worker",
  "version": "1.0.0",
  "description": "A worker for pulling movie data from a Google Sheet and the OMDb API",
  "main": "index.js",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1",
    "format": "prettier --write '**/*.{js,css,json,md}'",
    "build": "webpack --mode=production --env.runtime=webworker",
    "dev": "webpack --mode=development --watch --env.runtime=node",
    "analyze": "webpack --mode=development --watch --env.analyze --env.runtime=webworker"
  },
  "author": "Kevin Kipp <kevin.kipp@gmail.com>",
  "license": "MIT",
  "devDependencies": {
    "husky": "^2.1.0",
    "prettier": "^1.17.0",
    "webpack-bundle-analyzer": "^3.3.2"
  },
  "husky": {
    "hooks": {
      "pre-commit": "npm run format",
      "pre-push": "npm run format"
    }
  },
  "dependencies": {
    "@babel/core": "^7.5.4",
    "@babel/preset-env": "^7.5.4",
    "@babel/preset-typescript": "^7.3.3",
    "@types/node-fetch": "^2.3.7",
    "babel-loader": "^8.0.6",
    "dotenv": "^8.0.0",
    "json-loader": "^0.5.7",
    "jsrsasign": "^8.0.12",
    "node-fetch": "^2.6.0",
    "typescript": "^3.5.3",
    "webpack": "^4.35.3",
    "webpack-cli": "^3.3.5"
  }
}
```

It looks like supporting this would be a minor change here in `wranglerjs/index.js` and I'd be happy to open a PR! I'd also like to wrap the majority of this code in an `async` IIFE to `await` the promise of the config if there is one.

https://github.com/cloudflare/wrangler/blob/master/wranglerjs/index.js#L38

**Comment by third774** [comment]
Regarding `async/await` in `wranglerjs/index.js` — is there a minimum version of Node that must be supported?

**Comment by xtuc** [comment]
Thanks for your report.
I assumed that webpack will handle that configuration but it seems we have to do it on our side. It would be helpful if you can send a PR.

`async/await` is available since Node 7, that version is deprecated. I think we can use it, however I'm waiting for other thoughts.

**Comment by ashleymichal** [comment]
closed by #325
